streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -38,346 +39,678 @@
+        // NOTE we do not book keep the current offset since we leverage state manager as its source of truth
+        private Long restoreEndOffset;
+
+        // only for standby tasks that use source topics as changelogs (for active it is null);
+        // if it is not on source topics it is also null
+        private Long restoreLimitOffset;
+
+        // buffer records polled by the restore consumer;
+        private List<ConsumerRecord<byte[], byte[]>> bufferedRecords;
+
+        // the limit index (exclusive) inside the buffered records beyond which should not be used to restore
+        // either due to limit offset (standby) or committed end offset (active)
+        private int bufferedLimitIndex;
+
+        private ChangelogMetadata(final TopicPartition changelogPartition, final ProcessorStateManager stateManager) {
+            this.stateManager = stateManager;
+            this.changelogPartition = changelogPartition;
+            this.changelogState = ChangelogState.REGISTERED;
+            this.restoreEndOffset = null;
+            this.totalRestored = 0L;
+
+            this.bufferedRecords = new ArrayList<>();
+            this.restoreLimitOffset = null;
+            this.bufferedLimitIndex = 0;
+        }
+
+        private void clear() {
+            this.bufferedRecords.clear();
+        }
+
+        @Override
+        public String toString() {
+            final Long currentOffset = stateManager.changelogOffsets().get(changelogPartition);
+            return changelogState + " " + stateManager.taskType() +
+                " (currentOffset + " + currentOffset + "endOffset " + restoreEndOffset + ", limitOffset " + restoreLimitOffset + ")";
+        }
+    }
+
+    private ChangelogReaderState state;
+
     private final Logger log;
-    private final Consumer<byte[], byte[]> restoreConsumer;
-    private final StateRestoreListener userStateRestoreListener;
-    private final Map<TopicPartition, Long> restoreToOffsets = new HashMap<>();
-    private final Map<String, List<PartitionInfo>> partitionInfo = new HashMap<>();
-    private final Map<TopicPartition, StateRestorer> stateRestorers = new HashMap<>();
-    private final Set<TopicPartition> needsRestoring = new HashSet<>();
-    private final Set<TopicPartition> needsInitializing = new HashSet<>();
-    private final Set<TopicPartition> completedRestorers = new HashSet<>();
     private final Duration pollTime;
 
-    public StoreChangelogReader(final Consumer<byte[], byte[]> restoreConsumer,
-                                final Duration pollTime,
-                                final StateRestoreListener userStateRestoreListener,
-                                final LogContext logContext) {
+    // 1) we keep adding partitions to restore consumer whenever new tasks are registered with the state manager;
+    // 2) we do not unassign partitions when we switch between standbys and actives, we just pause / resume them;
```

Comment:
Can you elaborate why pausing/resuming is better than assigning/unassigning of partitions? I don't see a big difference?

Reply:
Assigning / unassigning would cleanup the internal bookkeeping data as well as the fetched records while pausing / restoring would not, so it is a lighter approach for better performance. Also it is clearer logic since with this approach the assigned partitions always correspond to the current assigned tasks (both active and standby), previously we have to keep track of the assigned partitions inside AssignedTasks as well since this correspondence was not there, now we do not need to keep track of this as there's always an exact mapping.